xfce4-terminal: embed Hack Braille and make it the default font

Compile Hack Braille (Hack 3.003 plus the braille patterns U+2800..U+28FF,
four styles; MIT and Bitstream Vera License) into xfce4-terminal and
default font-name to "Hack Braille 12".  The fonts are registered with
fontconfig from memory (memfd) at startup; nothing is installed.  A font
set in the preferences still wins.  Adds a fontconfig dependency.

Needs hack-braille-3.003/ (HackBraille-*.ttf, LICENSE.md, README,
mkhackbraille.py) unpacked into the source tree.
Apply after xfce4-terminal-1.2-paste-warning-default-false.patch.

Commits (branch embedded-font on xfce4-terminal-1.1.5):
  5c76cf1 Embed Hack Braille and make it the default font

diff --git a/configure.ac b/configure.ac
index d475607..2fc32b5 100644
--- a/configure.ac
+++ b/configure.ac
@@ -17,6 +17,7 @@ m4_define([glib_min_version], [2.44.0])
 m4_define([gtk_min_version], [3.22.0])
 m4_define([vte_min_version], [0.51.3])
 m4_define([pcre2_min_version], [10.00])
+m4_define([fontconfig_min_version], [2.11.0])
 m4_define([libxfce4ui_min_version], [4.17.5])
 m4_define([libxfce4util_min_version], [4.16.0])
 m4_define([xfconf_min_version], [4.16.0])
@@ -122,6 +123,7 @@ XDT_CHECK_PACKAGE([GIO], [gio-2.0], [glib_min_version])
 XDT_CHECK_PACKAGE([GTK], [gtk+-3.0], [gtk_min_version])
 XDT_CHECK_PACKAGE([VTE], [vte-2.91], [vte_min_version])
 XDT_CHECK_PACKAGE([PCRE2], [libpcre2-8], [pcre2_min_version])
+XDT_CHECK_PACKAGE([FONTCONFIG], [fontconfig], [fontconfig_min_version])
 XDT_CHECK_PACKAGE([LIBXFCE4UI], [libxfce4ui-2], [libxfce4ui_min_version])
 XDT_CHECK_PACKAGE([LIBXFCE4UTIL], [libxfce4util-1.0], [libxfce4util_min_version])
 XDT_CHECK_PACKAGE([XFCONF], [libxfconf-0], [xfconf_min_version])
diff --git a/terminal/Makefile.am b/terminal/Makefile.am
index a1df6c9..9517538 100644
--- a/terminal/Makefile.am
+++ b/terminal/Makefile.am
@@ -37,6 +37,8 @@ xfce4_terminal_SOURCES = \
 	$(xfce4_terminal_headers) \
 	main.c \
 	terminal-app.c \
+	terminal-embedded-fonts.c \
+	terminal-embedded-fonts.h \
 	terminal-encoding-action.c \
 	terminal-gdbus.c \
 	terminal-image-loader.c \
@@ -61,6 +63,7 @@ xfce4_terminal_CFLAGS = \
 	$(LIBXFCE4KBD_PRIVATE_CFLAGS) \
 	$(XFCONF_CFLAGS) \
 	$(PCRE2_CFLAGS) \
+	$(FONTCONFIG_CFLAGS) \
 	$(PLATFORM_CFLAGS)
 
 xfce4_terminal_LDFLAGS = \
@@ -78,7 +81,20 @@ xfce4_terminal_LDADD = \
 	$(LIBXFCE4UTIL_LIBS) \
 	$(LIBXFCE4KBD_PRIVATE_LIBS) \
 	$(XFCONF_LIBS) \
-	$(PCRE2_LIBS)
+	$(PCRE2_LIBS) \
+	$(FONTCONFIG_LIBS)
+
+# fonts compiled into xfce4-terminal, see terminal-embedded-fonts.c
+EMBEDDED_FONTS = \
+	$(top_srcdir)/hack-braille-3.003/HackBraille-Regular.ttf \
+	$(top_srcdir)/hack-braille-3.003/HackBraille-Bold.ttf \
+	$(top_srcdir)/hack-braille-3.003/HackBraille-Italic.ttf \
+	$(top_srcdir)/hack-braille-3.003/HackBraille-BoldItalic.ttf
+
+embedded-fonts.h: $(EMBEDDED_FONTS) $(srcdir)/mkembedfont.sh
+	$(SHELL) $(srcdir)/mkembedfont.sh $(EMBEDDED_FONTS) > $@
+
+xfce4_terminal-terminal-embedded-fonts.$(OBJEXT): embedded-fonts.h
 
 if HAVE_UTEMPTER
 xfce4_terminal_LDADD += -lutempter
@@ -162,7 +178,11 @@ clean-local:
 EXTRA_DIST = \
 	terminal-config.h.in \
 	terminal-marshal.list \
+	mkembedfont.sh \
 	meson.build \
 	$(NULL)
 
+MOSTLYCLEANFILES = \
+	embedded-fonts.h
+
 # vi:set ts=8 sw=8 noet ai nocindent:
diff --git a/terminal/main.c b/terminal/main.c
index 9102724..ac36b79 100644
--- a/terminal/main.c
+++ b/terminal/main.c
@@ -35,6 +35,7 @@
 #endif
 
 #include "terminal-app.h"
+#include "terminal-embedded-fonts.h"
 #include "terminal-gdbus.h"
 #include "terminal-preferences-dialog.h"
 #include "terminal-private.h"
@@ -203,6 +204,9 @@ main (int argc, char **argv)
 
   g_set_application_name (_("Xfce Terminal"));
 
+  /* make the compiled-in default font available, before GTK uses fonts */
+  terminal_embedded_fonts_register ();
+
   /* initialize GTK: do this before our parsing so GTK parses its options first */
   gtk_init (&argc, &argv);
 
diff --git a/terminal/mkembedfont.sh b/terminal/mkembedfont.sh
new file mode 100755
index 0000000..c0bbf32
--- /dev/null
+++ b/terminal/mkembedfont.sh
@@ -0,0 +1,27 @@
+#!/bin/sh
+# Generate embedded-fonts.h, which terminal-embedded-fonts.c compiles into
+# xfce4-terminal, from the font files given as arguments.  Uses only od and sed.
+set -e
+
+echo "/* generated by mkembedfont.sh - do not edit */"
+n=0
+for font in "$@"; do
+	echo
+	echo "/* ${font##*/} */"
+	echo "static const unsigned char embedded_font_$n[] = {"
+	od -An -v -tx1 "$font" | sed -e 's/ *\([0-9a-f][0-9a-f]\)/0x\1,/g'
+	echo "};"
+	n=`expr $n + 1`
+done
+
+echo
+echo "static const struct {"
+echo "    const unsigned char *data;"
+echo "    size_t size;"
+echo "} embedded_fonts[] = {"
+n=0
+for font in "$@"; do
+	echo "    { embedded_font_$n, sizeof(embedded_font_$n) },"
+	n=`expr $n + 1`
+done
+echo "};"
diff --git a/terminal/terminal-embedded-fonts.c b/terminal/terminal-embedded-fonts.c
new file mode 100644
index 0000000..414fe05
--- /dev/null
+++ b/terminal/terminal-embedded-fonts.c
@@ -0,0 +1,81 @@
+/*-
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * Register the fonts compiled into xfce4-terminal (see mkembedfont.sh) with
+ * fontconfig, so that the default font works without installing them.
+ *
+ * fontconfig only adds application fonts from files, so each font is copied
+ * into an anonymous memory file (memfd) and added by its /proc/self/fd path.
+ * fontconfig then sees all of its styles, and Pango opens them through that
+ * path.  The descriptors stay open for the life of the process, and are not
+ * inherited by the shell.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE /* memfd_create */
+#endif
+
+#include "terminal-embedded-fonts.h"
+
+#include <fontconfig/fontconfig.h>
+#include <stdio.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+#include "embedded-fonts.h"
+
+
+
+void
+terminal_embedded_fonts_register (void)
+{
+  gsize n;
+
+  for (n = 0; n < G_N_ELEMENTS (embedded_fonts); ++n)
+    {
+      const unsigned char *data = embedded_fonts[n].data;
+      gsize left = embedded_fonts[n].size;
+      gchar path[40];
+      gint fd;
+
+      fd = memfd_create ("xfce4-terminal-font", MFD_CLOEXEC);
+      if (fd < 0)
+        {
+          g_warning ("Cannot create memory file for embedded font");
+          return;
+        }
+
+      while (left != 0)
+        {
+          gssize wrote = write (fd, data, left);
+          if (wrote <= 0)
+            break;
+          data += wrote;
+          left -= wrote;
+        }
+
+      g_snprintf (path, sizeof (path), "/proc/self/fd/%d", fd);
+      if (left != 0 || !FcConfigAppFontAddFile (NULL, (const FcChar8 *) path))
+        {
+          g_warning ("Cannot add embedded font %s", path);
+          close (fd);
+        }
+    }
+}
diff --git a/terminal/terminal-embedded-fonts.h b/terminal/terminal-embedded-fonts.h
new file mode 100644
index 0000000..00975dd
--- /dev/null
+++ b/terminal/terminal-embedded-fonts.h
@@ -0,0 +1,27 @@
+/*-
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef TERMINAL_EMBEDDED_FONTS_H
+#define TERMINAL_EMBEDDED_FONTS_H
+
+#include <glib.h>
+
+G_BEGIN_DECLS
+
+void terminal_embedded_fonts_register (void);
+
+G_END_DECLS
+
+#endif /* !TERMINAL_EMBEDDED_FONTS_H */
diff --git a/terminal/terminal-preferences.c b/terminal/terminal-preferences.c
index ef24c64..50247a3 100644
--- a/terminal/terminal-preferences.c
+++ b/terminal/terminal-preferences.c
@@ -805,7 +805,7 @@ terminal_preferences_class_init (TerminalPreferencesClass *klass)
     g_param_spec_string ("font-name",
                          NULL,
                          "FontName",
-                         "Monospace 12",
+                         "Hack Braille 12", /* compiled in, see terminal-embedded-fonts.c */
                          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS);
 
   /**
